packages/eslint-plugin/tests/rules/no-unsafe-assignment.test.ts

```diff
@@ -12,55 +12,51 @@ type Options = InferOptionsTypeFromRule<typeof rule>;
 type MessageIds = InferMessageIdsTypeFromRule<typeof rule>;
 type InvalidTest = InvalidTestCase<MessageIds, Options>;
 
-function assignmentTest(
+const assignmentTest = (
   tests: [string, number, number, boolean?][],
-): InvalidTest[] {
-  return tests.reduce<InvalidTest[]>(
-    (acc, [assignment, column, endColumn, skipAssignmentExpression]) => {
-      // VariableDeclaration
-      acc.push({
-        code: `const ${assignment}`,
-        errors: [
-          {
-            messageId: 'unsafeArrayPatternFromTuple',
-            line: 1,
-            column: column + 6,
-            endColumn: endColumn + 6,
-          },
-        ],
-      });
-      // AssignmentPattern
-      acc.push({
-        code: `function foo(${assignment}) {}`,
-        errors: [
+): InvalidTest[] =>
+  tests.flatMap(([assignment, column, endColumn, skipAssignmentExpression]) => [
+    // VariableDeclaration
+    {
+      code: `const ${assignment}`,
+      errors: [
+        {
+          messageId: 'unsafeArrayPatternFromTuple',
+          line: 1,
+          column: column + 6,
+          endColumn: endColumn + 6,
+        },
+      ],
+    },
+    // AssignmentPattern
+    {
+      code: `function foo(${assignment}) {}`,
+      errors: [
+        {
+          messageId: 'unsafeArrayPatternFromTuple',
+          line: 1,
+          column: column + 13,
+          endColumn: endColumn + 13,
+        },
+      ],
+    },
+    // AssignmentExpression
+    ...(skipAssignmentExpression
```

Comment:
still think these types of cases should be explicit like below rather than using lots of `...` operator to conditionally push within an expression.
```ts
const mappedTests = [ 
  // ... lots of cases
]
if (skipAssignmentExpression !== true) {
    mappedTests.push(/* etc */);
}
```
(also note that `skipAssignmentExpression !== true` isn't equivalent to `if (!skipAssignmentExpression)` here - let's preserve that).